### DISCUSSION TITLE
HDMI CEC support on Xavier NX

### DISCUSSION THREAD

**Discussion by brgl** [OE4T/meta-tegra#558] [General]
Hi! I'd like to read HDMI CEC commands on the Xavier NX. Preferably routed to linux's input device subsystem but the tegra v4.9 kernel may be too old for that.

I'm seeing a misc driver in the tegra linux source under `nvidia/drivers/misc/tegra-cec/`. It registers a character device under `/dev/tegra_cec` and exposes a set of `ioctl()s`:

```
#define TEGRA_CEC_IOCTL_ERROR_RECOVERY          _IO(TEGRA_CEC_IOC_MAGIC, 1)
#define TEGRA_CEC_IOCTL_DUMP_REGISTERS          _IO(TEGRA_CEC_IOC_MAGIC, 2)
#define TEGRA_CEC_IOCTL_SET_RX_SNOOP            _IO(TEGRA_CEC_IOC_MAGIC, 3)
#define TEGRA_CEC_IOCTL_GET_RX_SNOOP            _IO(TEGRA_CEC_IOC_MAGIC, 4)
#define TEGRA_CEC_IOCTL_GET_POST_RECOVERY       _IO(TEGRA_CEC_IOC_MAGIC, 5)
```

Unfortunately `libcec` doesn't seem to support it and just from looking at the code I'm not quite sure how to use if I were to write the code myself.

In mainline linux, there's a proper tegra CEC driver and the device-tree for tegra194 defines a CEC device (although with a non-existent compatible - I can just assume the upstream driver will support this controller too if the compatible is added). However if I use meta-tegra with mainline linux, I can't boot to user-space correctly - systemd goes into emergency mode. I guess this is because the regular arm64 defconfig is not enough?

Has anyone had any luck either with the out-of-tree CEC driver or with correctly booting the mainline linux in yocto with system?

**Comment by kekiefer** [comment]
Regarding the mainline kernel, this is something a few of us have been wanting to try but haven't yet. I guess you've seen [the uefi boot firmware](https://developer.download.nvidia.com/embedded/L4T/UEFI_Readme.html) from nvidia? Based on this doc it looks like they are using the standard arm64 defconfig with some additional CMA and a few additional ethernet options. It would also seem you can choose configuration via ACPI or device tree, so there may be some hints in the provided dtb for AGX that would be applicable to NX

**Reply by brgl** [reply]
Thanks, I'll try to tweak my defconfig then. I think I'm pretty close to booting mainline though - systemd may be missing some kernel features but I'm pretty sure a simple busybox userspace will boot just fine - I'll try to do it on Monday.

**Reply by brgl** [reply]
@kekiefer FYI simple busybox user-space boots just fine with the [following](https://github.com/brgl/meta-tegra/tree/topic/mainline-kernel) quick & dirty patch.
